File: src/otogo/guard.py
"""Authority enforcement: the loop may change the metal, not the stone.

Three levels, per the design:
  free    - normal implementation levers, changed freely inside scope
  propose - crosses a boundary; the loop records evidence, a person decides
  frozen  - the stone itself (corpus, fixture, score rules, rubric, config)

Plus one rule that is not a level: measure files are APPEND-ONLY. The loop may
add a regression check for a failure it observed; it may not weaken or delete
an existing one, because that would let it grade its own repair.
"""
from __future__ import annotations

import difflib
import json
import shutil
from dataclasses import dataclass, field
from pathlib import Path

from .core import GlobSet, Loop, sha256_file, walk_repo


@dataclass
class Snapshot:
    hashes: dict          # rel -> sha256
    measure_files: list   # rels copied verbatim into the round dir
    config: dict          # authority AS OF round start — see check()

    def to_dict(self):
        return {"hashes": self.hashes, "measure_files": self.measure_files,
                "config": self.config}


def take_snapshot(loop: Loop, round_dir: Path) -> Snapshot:
    """Freeze the stone at round start so `guard` has something to compare to."""
    frozen, measure = loop.frozen(), loop.measure()
    hashes, measure_files = {}, []
    copy_root = round_dir / "_snapshot"
    for rel, path in walk_repo(loop.root):
        # Hash everything: `guard` reports a real diff, not just drift in the stone.
        hashes[rel] = sha256_file(path)
        if measure.match(rel):
            measure_files.append(rel)
            dest = copy_root / rel
            dest.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(path, dest)
    # The authority is captured with the hashes. check() must judge the round by
    # the rules in force when it opened, never by rules the round itself edited:
    # a loop that can empty its own `frozen` list has no authority model at all.
    snap = Snapshot(hashes, sorted(measure_files),
                    {"authority": loop.config["authority"],
                     "measure": loop.config["measure"]})
    (round_dir / "snapshot.json").write_text(json.dumps(snap.to_dict(), indent=2) + "\n")
    return snap


@dataclass
class Verdict:
    violations: list = field(default_factory=list)   # hard stops
    proposals: list = field(default_factory=list)    # need a human decision
    warnings: list = field(default_factory=list)     # attribution smells
    changed_free: list = field(default_factory=list)
    appended_measures: list = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return not self.violations

    def to_dict(self):
        return {"ok": self.ok, "violations": self.violations, "proposals": self.proposals,
                "warnings": self.warnings, "changed_free": self.changed_free,
                "appended_measures": self.appended_measures}


def _json_only_grew(old, new, path="$"):
    """Structural append-only for JSON evidence ledgers.

    Line diffs are the wrong tool for a file like docs/witnesses.json: adding a
    witness to an existing claim edits the previous line's comma, which reads as
    tampering when it is the opposite. So compare the parsed values instead —
    keys and array elements may be ADDED, never removed, reordered, or altered.
    Returns a list of violation strings; empty means the ledger only grew.
    """
    bad = []
    if isinstance(old, dict):
        if not isinstance(new, dict):
            return [f"{path}: object replaced by {type(new).__name__}"]
        for k, v in old.items():
            if k not in new:
                return bad + [f"{path}.{k}: removed"]
            bad += _json_only_grew(v, new[k], f"{path}.{k}")
    elif isinstance(old, list):
        if not isinstance(new, list):
            return [f"{path}: array replaced by {type(new).__name__}"]
        if len(new) < len(old):
            return [f"{path}: {len(old) - len(new)} element(s) removed"]
        for i, v in enumerate(old):
            # Existing entries must survive in place; new ones append after.
            bad += _json_only_grew(v, new[i], f"{path}[{i}]")
    elif old != new:
        bad.append(f"{path}: {old!r} -> {new!r}")
    return bad


def _is_pure_append(old: str, new: str) -> bool:
    """True when every original line survives, in order, with only insertions."""
    old_lines, new_lines = old.splitlines(), new.splitlines()
    sm = difflib.SequenceMatcher(None, old_lines, new_lines, autojunk=False)
    return all(tag in ("equal", "insert") for tag, *_ in sm.get_opcodes())


def check(loop: Loop, round_dir: Path) -> Verdict:
    snap_path = round_dir / "snapshot.json"
    if not snap_path.is_file():
        raise FileNotFoundError(snap_path)
    snap = json.loads(snap_path.read_text())
    old_hashes: dict = snap["hashes"]

    # Authority as of `open`, not as it stands now.
    cfg = snap.get("config") or {"authority": loop.config["authority"],
                                 "measure": loop.config["measure"]}
    frozen = GlobSet(cfg["authority"]["frozen"])
    propose = GlobSet(cfg["authority"]["propose"])
    measure = GlobSet(cfg["measure"]["globs"])
    json_ledgers = GlobSet(cfg["measure"].get("json_ledgers") or [])
    append_only = cfg["measure"].get("append_only", True)
    v = Verdict()
    seen = set()

    for rel, path in walk_repo(loop.root):
        seen.add(rel)
        digest = sha256_file(path)
        was = old_hashes.get(rel)
        changed = was is None or was != digest
        fpat = frozen.match(rel)
        mpat = measure.match(rel)
        ppat = propose.match(rel)

        if fpat and changed:
            kind = "added to" if was is None else "modified"
            v.violations.append(
                f"FROZEN {kind}: {rel} (matches `{fpat}`) — this file is part of the stone.")
            continue

        if mpat and changed:
            if was is None:
                v.appended_measures.append(f"{rel} (new check file)")
            elif json_ledgers.match(rel):
                old_text = (round_dir / "_snapshot" / rel).read_text(errors="replace")
                new_text = path.read_text(errors="replace")
                try:
                    diffs = _json_only_grew(json.loads(old_text), json.loads(new_text))
                except json.JSONDecodeError as e:
                    v.violations.append(f"LEDGER unparseable: {rel} — {e}")
                    continue
                if diffs:
                    v.violations.append(
                        f"LEDGER weakened: {rel} — " + "; ".join(diffs[:4])
                        + ("; ..." if len(diffs) > 4 else "")
                        + ". Evidence ledgers may gain entries, never lose or rewrite them.")
                else:
                    v.appended_measures.append(f"{rel} (ledger grew)")
            elif append_only:
                old_text = (round_dir / "_snapshot" / rel).read_text(errors="replace")
                new_text = path.read_text(errors="replace")
                if _is_pure_append(old_text, new_text):
                    v.appended_measures.append(f"{rel} (appended)")
                else:
                    v.violations.append(
                        f"MEASURE weakened: {rel} — existing lines were changed or removed. "
                        "Checks are append-only; a round may add a rung, never lower one.")
            else:
                v.appended_measures.append(f"{rel} (modified)")
            continue

        if ppat and changed:
            v.proposals.append(f"{rel} (matches `{ppat}`)")
            continue

        if changed:
            v.changed_free.append(rel)

    for rel in old_hashes:
        if rel in seen:
            continue
        if frozen.match(rel):
            v.violations.append(f"FROZEN deleted: {rel} — the stone cannot shrink.")
        elif measure.match(rel):
            v.violations.append(f"MEASURE deleted: {rel} — checks are append-only.")
        elif propose.match(rel):
            v.proposals.append(f"{rel} (deleted)")
        else:
            v.changed_free.append(f"{rel} (deleted)")

    # One gap does not mean one file, but it does mean one causal claim.
    limit = loop.config["budget"]["attribution_warn_areas"]
    areas = {r.split("/")[0] for r in v.changed_free}
    if len(areas) > limit:
        v.warnings.append(
            f"attribution: this round touched {len(areas)} top-level areas "
            f"({', '.join(sorted(areas))}). If the round changes several independent levers, "
            "no one can tell which lesson to preserve.")
    if v.changed_free and not v.appended_measures:
        v.warnings.append(
            "no deterministic check was added. A closed gap without a new rung "
            "will be re-earned and re-lost.")
    return v

File: src/guard.rs
//! Authority enforcement: the loop may change the product, not the exam.
//!
//!   free    — implementation levers, changed freely inside scope
//!   propose — crosses a boundary; recorded, a person decides
//!   frozen  — the exam: corpus, fixture, score rules, rubric, loop.json
//!
//! Plus one rule that is not a level: measures are APPEND-ONLY. A round may add
//! a rung for a failure it observed; it may not weaken or delete one, because
//! that would let it grade its own repair.

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use serde_json::{json, Value};

use crate::core::{pretty, sha256_file, walk_repo, GoalLoop, LoopError, Res};
use crate::glob::GlobSet;

pub struct Verdict {
    pub violations: Vec<String>,
    pub proposals: Vec<String>,
    pub warnings: Vec<String>,
    pub changed_free: Vec<String>,
    pub appended_measures: Vec<String>,
}

impl Verdict {
    pub fn ok(&self) -> bool {
        self.violations.is_empty()
    }

    pub fn to_json(&self) -> Value {
        json!({
            "ok": self.ok(),
            "violations": self.violations,
            "proposals": self.proposals,
            "warnings": self.warnings,
            "changed_free": self.changed_free,
            "appended_measures": self.appended_measures,
        })
    }
}

/// Freeze the exam at round start. The authority is captured alongside the
/// hashes: a round must be judged by the rules in force when it opened, never
/// by rules the round itself edited — a loop that can empty its own `frozen`
/// list has no authority model at all.
pub fn take_snapshot(l: &GoalLoop, round_dir: &Path) -> Res<usize> {
    let measure = l.measure();
    let mut hashes = BTreeMap::new();
    let mut measure_files = Vec::new();
    let copy_root = round_dir.join("_snapshot");

    for (rel, path) in walk_repo(&l.root) {
        hashes.insert(rel.clone(), sha256_file(&path)?);
        if measure.is_match(&rel) {
            measure_files.push(rel.clone());
            let dest = copy_root.join(&rel);
            if let Some(p) = dest.parent() {
                fs::create_dir_all(p).ok();
            }
            fs::copy(&path, &dest).map_err(|e| LoopError(format!("{rel}: {e}")))?;
        }
    }
    let n = hashes.len();
    let snap = json!({
        "hashes": hashes,
        "measure_files": measure_files,
        "config": {
            "authority": l.config.get("authority").cloned().unwrap_or(Value::Null),
            "measure": l.config.get("measure").cloned().unwrap_or(Value::Null),
        }
    });
    fs::write(
        round_dir.join("snapshot.json"),
        format!("{}\n", pretty(&snap)),
    )
    .map_err(|e| LoopError(e.to_string()))?;
    Ok(n)
}

/// True when every original line survives, in order, with only insertions —
/// i.e. `old` is a subsequence of `new`.
fn is_pure_append(old: &str, new: &str) -> bool {
    let mut it = new.lines();
    for line in old.lines() {
        let mut found = false;
        for candidate in it.by_ref() {
            if candidate == line {
                found = true;
                break;
            }
        }
        if !found {
            return false;
        }
    }
    true
}

/// Structural append-only for JSON evidence ledgers.
///
/// A line diff is the wrong tool here: adding a witness to an existing claim
/// edits the previous line's comma, which reads as tampering when it is the
/// opposite. Compare parsed values instead — keys and array elements may be
/// ADDED, never removed, reordered, or altered.
fn json_only_grew(old: &Value, new: &Value, path: &str, bad: &mut Vec<String>) {
    match (old, new) {
        (Value::Object(o), Value::Object(n)) => {
            for (k, v) in o {
                match n.get(k) {
                    Some(nv) => json_only_grew(v, nv, &format!("{path}.{k}"), bad),
                    None => bad.push(format!("{path}.{k}: removed")),
                }
            }
        }
        (Value::Object(_), other) => {
            bad.push(format!("{path}: object replaced by {}", kind(other)))
        }
        (Value::Array(o), Value::Array(n)) => {
            if n.len() < o.len() {
                bad.push(format!("{path}: {} element(s) removed", o.len() - n.len()));
                return;
            }
            for (i, v) in o.iter().enumerate() {
                // Existing entries must survive in place; new ones append after.
                json_only_grew(v, &n[i], &format!("{path}[{i}]"), bad);
            }
        }
        (Value::Array(_), other) => bad.push(format!("{path}: array replaced by {}", kind(other))),
        (a, b) if a != b => bad.push(format!("{path}: {a} -> {b}")),
        _ => {}
    }
}

fn kind(v: &Value) -> &'static str {
    match v {
        Value::Null => "null",
        Value::Bool(_) => "bool",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

pub fn check(l: &GoalLoop, round_dir: &Path) -> Res<Verdict> {
    let snap_path = round_dir.join("snapshot.json");
    let raw = fs::read_to_string(&snap_path).map_err(|_| {
        LoopError(format!(
            "{} missing — was the round opened?",
            snap_path.display()
        ))
    })?;
    let snap: Value = serde_json::from_str(&raw).map_err(|e| LoopError(e.to_string()))?;
    let old_hashes = snap.get("hashes").cloned().unwrap_or(Value::Null);

    // Authority as of `open`, not as it stands now.
    let cfg = snap.get("config").cloned().unwrap_or_else(
        || json!({"authority": l.config.get("authority"), "measure": l.config.get("measure")}),
    );
    let frozen = GlobSet::from_json(cfg.pointer("/authority/frozen"));
    let propose = GlobSet::from_json(cfg.pointer("/authority/propose"));
    let measure = GlobSet::from_json(cfg.pointer("/measure/globs"));
    let ledgers = GlobSet::from_json(cfg.pointer("/measure/json_ledgers"));
    let append_only = cfg
        .pointer("/measure/append_only")
        .and_then(|v| v.as_bool())
        .unwrap_or(true);

    let mut v = Verdict {
        violations: vec![],
        proposals: vec![],
        warnings: vec![],
        changed_free: vec![],
        appended_measures: vec![],
    };
    let mut seen: Vec<String> = Vec::new();

    for (rel, path) in walk_repo(&l.root) {
        seen.push(rel.clone());
        let digest = sha256_file(&path)?;
        let was = old_hashes.get(&rel).and_then(|h| h.as_str());
        let changed = was.map(|w| w != digest).unwrap_or(true);

        if let Some(pat) = frozen.match_of(&rel) {
            if changed {
                let kind = if was.is_none() {
                    "added to"
                } else {
                    "modified"
                };
                v.violations.push(format!(
                    "FROZEN {kind}: {rel} (matches `{pat}`) — this file is part of the exam."
                ));
            }
            continue;
        }

        if measure.is_match(&rel) {
            if !changed {
                continue;
            }
            if was.is_none() {
                v.appended_measures.push(format!("{rel} (new check file)"));
            } else if ledgers.is_match(&rel) {
                let old_text =
                    fs::read_to_string(round_dir.join("_snapshot").join(&rel)).unwrap_or_default();
                let new_text = fs::read_to_string(&path).unwrap_or_default();
                match (
                    serde_json::from_str::<Value>(&old_text),
                    serde_json::from_str::<Value>(&new_text),
                ) {
                    (Ok(o), Ok(n)) => {
                        let mut bad = Vec::new();
                        json_only_grew(&o, &n, "$", &mut bad);
                        if bad.is_empty() {
                            v.appended_measures.push(format!("{rel} (ledger grew)"));
                        } else {
                            let shown: Vec<String> = bad.iter().take(4).cloned().collect();
                            v.violations.push(format!(
                                "LEDGER weakened: {rel} — {}{}. Evidence ledgers may gain \
                                 entries, never lose or rewrite them.",
                                shown.join("; "),
                                if bad.len() > 4 { "; ..." } else { "" }
                            ));
                        }
                    }
                    _ => v.violations.push(format!("LEDGER unparseable: {rel}")),
                }
            } else if append_only {
                let old_text =
                    fs::read_to_string(round_dir.join("_snapshot").join(&rel)).unwrap_or_default();
                let new_text = fs::read_to_string(&path).unwrap_or_default();
                if is_pure_append(&old_text, &new_text) {
                    v.appended_measures.push(format!("{rel} (appended)"));
                } else {
                    v.violations.push(format!(
                        "MEASURE weakened: {rel} — existing lines were changed or removed. \
                         Checks are append-only; a round may add a rung, never lower one."
                    ));
                }
            } else {
                v.appended_measures.push(format!("{rel} (modified)"));
            }
            continue;
        }

        if let Some(pat) = propose.match_of(&rel) {
            if changed {
                v.proposals.push(format!("{rel} (matches `{pat}`)"));
            }
            continue;
        }

        if changed {
            v.changed_free.push(rel);
        }
    }

    if let Some(map) = old_hashes.as_object() {
        for rel in map.keys() {
            if seen.iter().any(|s| s == rel) {
                continue;
            }
            if frozen.is_match(rel) {
                v.violations
                    .push(format!("FROZEN deleted: {rel} — the exam cannot shrink."));
            } else if measure.is_match(rel) {
                v.violations
                    .push(format!("MEASURE deleted: {rel} — checks are append-only."));
            } else if propose.is_match(rel) {
                v.proposals.push(format!("{rel} (deleted)"));
            } else {
                v.changed_free.push(format!("{rel} (deleted)"));
            }
        }
    }

    // One gap does not mean one file, but it does mean one causal claim.
    let limit = l
        .config
        .pointer("/budget/attribution_warn_areas")
        .and_then(|x| x.as_u64())
        .unwrap_or(3) as usize;
    let mut areas: Vec<&str> = v
        .changed_free
        .iter()
        .map(|r| r.split('/').next().unwrap_or(r))
        .collect();
    areas.sort_unstable();
    areas.dedup();
    if areas.len() > limit {
        v.warnings.push(format!(
            "attribution: this round touched {} top-level areas ({}). If the round changes \
             several independent levers, no one can tell which lesson to preserve.",
            areas.len(),
            areas.join(", ")
        ));
    }
    if !v.changed_free.is_empty() && v.appended_measures.is_empty() {
        v.warnings.push(
            "no deterministic check was added. A closed gap without a new rung will be \
             re-earned and re-lost."
                .into(),
        );
    }
    Ok(v)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn append_is_allowed_and_edits_are_not() {
        assert!(is_pure_append("a\nb\n", "a\nb\nc\n"));
        assert!(is_pure_append("a\nb\n", "x\na\ny\nb\nz\n"));
        assert!(!is_pure_append("a\nb\n", "a\n"));
        assert!(!is_pure_append("assert x == 2\n", "assert True\n"));
    }

    fn grew(o: Value, n: Value) -> Vec<String> {
        let mut bad = Vec::new();
        json_only_grew(&o, &n, "$", &mut bad);
        bad
    }

    #[test]
    fn a_ledger_may_gain_a_witness_or_a_claim() {
        let base = json!({"c": {"claim": "x", "witnesses": ["go:A"]}});
        assert!(grew(
            base.clone(),
            json!({"c": {"claim": "x", "witnesses": ["go:A", "sdk:B"]}})
        )
        .is_empty());
        assert!(grew(
            base,
            json!({"c": {"claim": "x", "witnesses": ["go:A"]}, "d": {"claim": "y"}})
        )
        .is_empty());
    }

    #[test]
    fn a_ledger_may_not_lose_or_rewrite_evidence() {
        let base = json!({"c": {"claim": "x", "witnesses": ["go:A", "go:B"]}});
        assert!(!grew(
            base.clone(),
            json!({"c": {"claim": "x", "witnesses": ["go:A"]}})
        )
        .is_empty());
        assert!(!grew(
            base.clone(),
            json!({"c": {"claim": "x", "witnesses": ["go:A", "go:Easier"]}})
        )
        .is_empty());
        assert!(!grew(
            base.clone(),
            json!({"c": {"claim": "softer", "witnesses": ["go:A", "go:B"]}})
        )
        .is_empty());
        assert!(!grew(base, json!({})).is_empty());
    }
}
